File: android/src/main/java/com/nativesprings/shaders/CloudsOverlay.kt
package com.nativesprings.shaders

import android.content.Context
import android.opengl.GLES30

class CloudsOverlay(private val context: Context) : Overlay {
    override val name: String = "clouds"

    override val parameters: List<ShaderParameter> = listOf(
        ShaderParameter("intensity", ShaderParameterType.FLOAT, 0.6f),
        ShaderParameter("color", ShaderParameterType.FLOAT3, listOf(0.15f, 0.2f, 0.55f)),
        ShaderParameter("secondaryColor", ShaderParameterType.FLOAT3, listOf(0.25f, 0.15f, 0.5f)),
        ShaderParameter("tertiaryColor", ShaderParameterType.FLOAT3, listOf(0.1f, 0.3f, 0.5f)),
        ShaderParameter("scale", ShaderParameterType.FLOAT, 1.0f),
        ShaderParameter("speed", ShaderParameterType.FLOAT, 0.3f),
        ShaderParameter("softness", ShaderParameterType.FLOAT, 0.6f),
        ShaderParameter("blobCount", ShaderParameterType.FLOAT, 3.0f)
    )

    override val needsAnimation: Boolean = true

    private var time: Float = 0.0f

    // Cached uniform locations — populated once in compile(), reused every frame.
    private var locTime = -1
    private var locIntensity = -1
    private var locViewSize = -1
    private var locColor = -1
    private var locSecondaryColor = -1
    private var locTertiaryColor = -1
    private var locScale = -1
    private var locSpeed = -1
    private var locSoftness = -1
    private var locBlobCount = -1

    init {
        OverlayRegistry.register(this)
    }

    override fun compile(): Int {
        val vertexShader = GLUtils.loadShaderFromResource(context, R.raw.vertex)
        val fragmentShader = GLUtils.loadShaderFromResource(context, R.raw.clouds)
        val programId = GLUtils.createProgram(vertexShader, fragmentShader)
        if (programId != 0) {
            locTime           = GLES30.glGetUniformLocation(programId, "time")
            locIntensity      = GLES30.glGetUniformLocation(programId, "intensity")
            locViewSize       = GLES30.glGetUniformLocation(programId, "viewSize")
            locColor          = GLES30.glGetUniformLocation(programId, "color")
            locSecondaryColor = GLES30.glGetUniformLocation(programId, "secondaryColor")
            locTertiaryColor  = GLES30.glGetUniformLocation(programId, "tertiaryColor")
            locScale          = GLES30.glGetUniformLocation(programId, "scale")
            locSpeed          = GLES30.glGetUniformLocation(programId, "speed")
            locSoftness       = GLES30.glGetUniformLocation(programId, "softness")
            locBlobCount      = GLES30.glGetUniformLocation(programId, "blobCount")
        }
        return programId
    }

    override fun update(deltaTime: Double) {
        time += deltaTime.toFloat()
    }

    override fun encode(programId: Int, context: OverlayContext) {
        if (programId == 0) {
            android.util.Log.e(TAG, "Invalid program ID (0) - shader compilation likely failed")
            return
        }

        GLES30.glUniform1f(locTime, time)

        val intensity = (context.parameters["intensity"] as? Number)?.toFloat() ?: 0.6f
        GLES30.glUniform1f(locIntensity, intensity)

        GLES30.glUniform2f(locViewSize, context.viewWidth.toFloat(), context.viewHeight.toFloat())

        val color = context.parameters.extractFloat3(
            "color", "colorR", "colorG", "colorB",
            floatArrayOf(0.15f, 0.2f, 0.55f)
        )
        GLES30.glUniform3f(locColor, color[0], color[1], color[2])

        val secondaryColor = context.parameters.extractFloat3(
            "secondaryColor", "secondaryColorR", "secondaryColorG", "secondaryColorB",
            floatArrayOf(0.25f, 0.15f, 0.5f)
        )
        GLES30.glUniform3f(locSecondaryColor, secondaryColor[0], secondaryColor[1], secondaryColor[2])

        val tertiaryColor = context.parameters.extractFloat3(
            "tertiaryColor", "tertiaryColorR", "tertiaryColorG", "tertiaryColorB",
            floatArrayOf(0.1f, 0.3f, 0.5f)
        )
        GLES30.glUniform3f(locTertiaryColor, tertiaryColor[0], tertiaryColor[1], tertiaryColor[2])

        val scale = (context.parameters["scale"] as? Number)?.toFloat() ?: 1.0f
        GLES30.glUniform1f(locScale, scale)

        val speed = (context.parameters["speed"] as? Number)?.toFloat() ?: 0.3f
        GLES30.glUniform1f(locSpeed, speed)

        val softness = (context.parameters["softness"] as? Number)?.toFloat() ?: 0.6f
        GLES30.glUniform1f(locSoftness, softness)

        val blobCount = (context.parameters["blobCount"] as? Number)?.toFloat() ?: 3.0f
        GLES30.glUniform1f(locBlobCount, blobCount)
    }

    companion object {
        private const val TAG = "CloudsOverlay"
    }
}

File: ios/CloudsOverlay.swift
import Foundation
import Metal

public class CloudsOverlay: Overlay {
    public var name: String { "clouds" }

    public var parameters: [ShaderParameter] {
        [
            ShaderParameter(name: "intensity", type: .float, defaultValue: 0.6),
            ShaderParameter(name: "color", type: .float3, defaultValue: [0.15, 0.2, 0.55]),
            ShaderParameter(name: "secondaryColor", type: .float3, defaultValue: [0.25, 0.15, 0.5]),
            ShaderParameter(name: "tertiaryColor", type: .float3, defaultValue: [0.1, 0.3, 0.5]),
            ShaderParameter(name: "scale", type: .float, defaultValue: 1.0),
            ShaderParameter(name: "speed", type: .float, defaultValue: 0.3),
            ShaderParameter(name: "softness", type: .float, defaultValue: 0.6),
            ShaderParameter(name: "blobCount", type: .float, defaultValue: 3.0),
        ]
    }

    public var needsAnimation: Bool { true }

    private var time: Float = 0.0

    public init() {
        OverlayRegistry.shared.register(self)
    }

    public func compile(device: MTLDevice) throws -> MTLRenderPipelineState {
        return try MetalLibraryLoader.loadAndCompilePipeline(
            resourceName: "Clouds",
            subdirectory: "Overlays",
            vertexFunctionName: "fullscreenVertex",
            fragmentFunctionName: "cloudsFragment",
            bundle: Bundle(for: type(of: self)),
            device: device,
            enableBlending: true
        )
    }

    public func update(deltaTime: TimeInterval) {
        time += Float(deltaTime)
    }

    public func encode(encoder: MTLRenderCommandEncoder, context: OverlayContext) {
        struct CloudsParameters {
            var time: Float
            var intensity: Float
            var viewSize: SIMD2<Float>
            var color: SIMD3<Float>
            var secondaryColor: SIMD3<Float>
            var tertiaryColor: SIMD3<Float>
            var scale: Float
            var speed: Float
            var softness: Float
            var blobCount: Float
        }

        let color = extractFloat3(
            from: context.parameters,
            arrayName: "color",
            rName: "colorR",
            gName: "colorG",
            bName: "colorB",
            defaultValue: SIMD3<Float>(0.15, 0.2, 0.55)
        )

        let secondaryColor = extractFloat3(
            from: context.parameters,
            arrayName: "secondaryColor",
            rName: "secondaryColorR",
            gName: "secondaryColorG",
            bName: "secondaryColorB",
            defaultValue: SIMD3<Float>(0.25, 0.15, 0.5)
        )

        let tertiaryColor = extractFloat3(
            from: context.parameters,
            arrayName: "tertiaryColor",
            rName: "tertiaryColorR",
            gName: "tertiaryColorG",
            bName: "tertiaryColorB",
            defaultValue: SIMD3<Float>(0.1, 0.3, 0.5)
        )

        var params = CloudsParameters(
            time: time,
            intensity: (context.parameters["intensity"] as? Float) ?? 0.6,
            viewSize: SIMD2<Float>(Float(context.viewSize.width), Float(context.viewSize.height)),
            color: color,
            secondaryColor: secondaryColor,
            tertiaryColor: tertiaryColor,
            scale: (context.parameters["scale"] as? Float) ?? 1.0,
            speed: (context.parameters["speed"] as? Float) ?? 0.3,
            softness: (context.parameters["softness"] as? Float) ?? 0.6,
            blobCount: (context.parameters["blobCount"] as? Float) ?? 3.0
        )

        encoder.setFragmentBytes(&params, length: MemoryLayout<CloudsParameters>.stride, index: 0)
        encoder.drawPrimitives(type: .triangle, vertexStart: 0, vertexCount: 6)
    }
}
